### DISCUSSION TITLE
Pip install error

### DISCUSSION THREAD

**Discussion by Grace-Chen-12** [RosettaCommons/rosetta#189] [PyRosetta] [answered]
I was trying to install a PyRosetta net in my Ubuntu on Windows 11 box, and I was using these two commands:
pip install pyrosetta-installer 
and
pip install /mnt/c/pyrosetta-2024.39+release.59628fbc5b-cp312-cp312-linux_x86_64.whl

I got the same error for both:
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.

    See /usr/share/doc/python3.12/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.

Can anybody tell me how to fix this? I ran the command "sudo apt install python3 python3-pip ipython3, and it seems I already have the latest Python version installed.

**Comment by roccomoretti** [comment, marked as answer]
You're using the system python. Due to "dependency hell" (technical term), certain distributions like Debian & Ubuntu turn off pip for the system python. The issue is that installing arbitrary packages can cause other packages to update/change version, which may potentially break the python scripts which are used by the OS. To prevent that, the distributions carefully control which packages/versions can be installed in the system python and do so by turning off pip and requiring you to install python packages for the system python through the distribution's standard package manager (e.g. `apt`).

The issue you run into is that neither Ubuntu nor Debian really have an `apt install`able PyRosetta. What you need to do instead is to create a personal virtual environment which is independent of the system python environment and install PyRosetta into that.

The recommended way to do that would be to install [Conda](https://docs.anaconda.com/miniconda/). Conda creates a Python installation in your home directory, and provides a way to easily switch between different python environments. I recommend Conda because it's the current _de facto_ way of installing many of the biochemistry-related packages in Python. (e.g. most of the ML packages have Conda install instructions).  -- As an added benefit, there's a RosettaCommons Conda channel where you can install PyRosetta through Conda directly (see the [PyRosetta downloads page](https://www.pyrosetta.org/downloads#h.c0px19b8kvuw) for more).

**Comment by lyskov** [comment]
I would voice somewhat different perspective on this: using Conda while convenient for simple projects frequently lead to rather heavy installs with a lot of unnecessary dependencies for complex projects. In particular, when installing PyRosetta from Conda using Python-3.9 or lower that install alone will pull ~1GB of other packages that you might not necessary need. So in some cases it might be best to keep things simple and go with creating Python virtual environment. And if you have to stick with Conda then i would recommend to use Python-3.10 or later. What approach will work for particular case will be project dependent so just keep an open mind on this and see what works.
